subprojects/dependency-management/src/test/groovy/org/gradle/internal/component/model/AttributeConfigurationSelectorTest.groovy

```diff
@@ -0,0 +1,250 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.component.model
+
+import com.google.common.base.Optional
+import com.google.common.collect.ImmutableList
+import org.gradle.api.artifacts.ModuleVersionIdentifier
+import org.gradle.api.artifacts.component.ComponentIdentifier
+import org.gradle.api.capabilities.CapabilitiesMetadata
+import org.gradle.api.capabilities.Capability
+import org.gradle.api.internal.attributes.AttributesSchemaInternal
+import org.gradle.api.internal.attributes.DefaultAttributesSchema
+import org.gradle.api.internal.attributes.ImmutableAttributes
+import org.gradle.internal.component.AmbiguousConfigurationSelectionException
+import org.gradle.internal.component.NoMatchingConfigurationSelectionException
+import org.gradle.util.SnapshotTestUtil
+import org.gradle.util.TestUtil
+import spock.lang.Specification
+import spock.lang.Unroll
+
+import static org.gradle.util.AttributeTestUtil.attributes
+
+class AttributeConfigurationSelectorTest extends Specification {
+    private final AttributesSchemaInternal attributesSchema = new DefaultAttributesSchema(new ComponentAttributeMatcher(), TestUtil.instantiatorFactory(), SnapshotTestUtil.valueSnapshotter())
+
+    private ComponentResolveMetadata targetComponent
+    private ConfigurationMetadata selected
+    private ImmutableAttributes consumerAttributes = ImmutableAttributes.EMPTY
+    private List<Capability> requestedCapabilities = []
+
+    @Unroll
+    def "selects a variant when there's no ambiguity"() {
+        given:
+        component(
+                variant("api", attributes('org.gradle.usage': 'java-api')),
+                variant("runtime", attributes('org.gradle.usage': 'java-runtime'))
+        )
+
+        and:
+        consumerAttributes('org.gradle.usage': usage)
+
+        when:
+        performSelection()
+
+        then:
+        selected.name == expected
+
+        where:
+        usage          | expected
+        'java-api'     | 'api'
+        'java-runtime' | 'runtime'
+    }
+
+    def "fails to select a variant when there are more than one candidate"() {
+        given:
+        component(
+                variant("api1", attributes('org.gradle.usage': 'java-api')),
+                variant("api2", attributes('org.gradle.usage': 'java-api'))
+        )
+
+        and:
+        consumerAttributes('org.gradle.usage': 'java-api')
+
+        when:
+        performSelection()
+
+        then:
+        AmbiguousConfigurationSelectionException e = thrown()
+        e.message == '''Cannot choose between the following variants of org:lib:1.0:
+  - api1
```

Comment:
I know that's not the goal of this PR, but this line and the next one feel like redudant information given that the variants are listed later anyway.
What about making the error message more compact?

Reply:
Let's do this later. We have a lot of tests and I don't want to spend time on fixing test error messages instead of fixing bugs.